feat: lint examples

naseemkullah opened the pull request:
## Which problem is this PR solving?
Fixes #603
-

## Short description of the changes

Adds .eslintrc at root of examples folder, lints said examples (WIP) and adds a step in pipeline to validate that the examples are linted.

naseemkullah commented:
> Each example has to have its own .eslintrc? is there no way to share a single one in the whole examples folder?

Sounds sensible, will look into that.
BTW `airbnb` tends to remove `'use strict;'`, do we care enough to override that rule? Should we try another eslint-config?

dyladan commented:
> > Each example has to have its own .eslintrc? is there no way to share a single one in the whole examples folder?
> 
> Sounds sensible, will look into that.
> BTW `airbnb` tends to remove `'use strict;'`, do we care enough to override that rule? Should we try another eslint-config?

I used the airbnb rules on a previous project and found them to be generally sensible, but one of the overrides we applied was to always use strict. The airbnb style guide assumes you are using the airbnb development toolchain which inserts `use strict` during the compile step :) https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/strict.js#L3

dyladan commented:
@naseemkullah it looks like one of the rules is no-use-before-defined whose behavior I don't happen to like. JS has function hoisting which I find makes code a lot easier to understand. I like to have the "main" function at the top, with the utility functions it calls at the bottom, which reads much more naturally to me. With the airbnb rules you are required to scroll past all the utility functions before you get to the function that actually does the work.

dyladan commented on examples/.eslintrc:
```diff
@@ -0,0 +1,13 @@
+  },
+  "rules": {
+    "strict": ["error", "global"],
+    "no-use-before-define": "off"
```
```suggestion
    "no-use-before-define": ["error", "nofunc"]
```

dyladan commented:
Can you add a yarn script `lint:examples` to the root package.json? Some people may not have eslint installed globally.

naseemkullah commented:
> Can you add a yarn script `lint:examples` to the root package.json? Some people may not have eslint installed globally.

Sure,
Why do some scripts seperate words with `:` and others with `-` ? What's the convention here?

dyladan commented:
> > Can you add a yarn script `lint:examples` to the root package.json? Some people may not have eslint installed globally.
> 
> Sure sounds good,
> Why do some scripts separate words with `:` (e.g. `codecov:browser`) and others with `-` (e.g. `docs-deploy`) ? What's the convention here?

no convention afaik. just different people adding them

dyladan commented on .circleci/config.yml:
```diff
@@ -173,7 +173,7 @@ jobs:
       - run:
           name: Lint examples
-          command: yarn run lint:examples
+          command: npm i --only=dev && npm run lint:examples
```
We are actually very intentionally not installing all of the dev dependencies here and only installing the bare minimum to speed up the time it takes the ci to run. That's why the global add on line 160 is there.

naseemkullah replied:
Hmm I thought that might be the case, just wanted to get something working first. I think we will need the eslint modules installed to run the lint though.

naseemkullah commented:
Below 504 is breaking the build
` [504] https://gitter.im/open-telemetry/opentelemetry-node?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge`

naseemkullah commented:
@dyladan @mayurkale22  Please provide thoughts on this commit which replaces the `setup` module with a `tracer` module that both initializes and exports the tracer, which is used in main `server` module. Please note that `http` can be imported/required before `tracer` provided that the server is started after `tracer` is imported/required.

Not only does this change fix various lint problems, it is a clearer way to implement tracing.

If you agree, I will go ahead and do something similar in rest of examples (and getting started guide, in a separate PR).

dyladan commented:
@naseemkullah i'm not sure what's up with your lint/docs but it is passing in other branches so something did change.

naseemkullah commented:
Please advise for:

`examples/grpc/helloworld_grpc_pb.js` and  `examples/grpc/helloworld_pb.js`  both have warnings that they are generated code and not to edit.

as well as:

```
/examples/redis/express-tracer-handlers.js
  24:27  error  '_next' is defined but never used  no-unused-vars
```

dyladan commented:
@markwolff generated that file so I would ask him. I don't see anything in the folder that gives away how he generated it.

dyladan commented:
As far as the _next goes, I would use `no-unused-vars: ["error", { "argsIgnorePattern": "^_" }]` and prefix any unused variables with an underscore like we do with privates and unused in the regular codebase.

naseemkullah commented:
Please advise for various:

```
`./tracer` import should occur after import of ...
```

caused by having tracer initialization before import of other modules.

naseemkullah commented:
@dyladan I've fixed conflicts and I've added many exceptions as there was too much non-linted code. Lets please get this in before more changes are made to examples to avoid fixing conflicts.

naseemkullah commented:
bump

dyladan commented on package.json:
```diff
@@ -23,7 +23,9 @@
     "docs": "lerna run docs",
-    "docs-deploy": "gh-pages --dist packages/opentelemetry-types/docs/out"
+    "docs-deploy": "gh-pages --dist packages/opentelemetry-types/docs/out",
+    "lint-examples": "node_modules/.bin/eslint ./examples/*/**.js",
```
you don't need to explicitly put `node_modules/.bin/eslint`, when run as part of an npm script this is done automatically.

dyladan replied:
You have your globbing backwards. It should be `./examples/**/*.js`

naseemkullah replied:
Oh good to know and oops yes thank you.

mayurkale22 commented:
I think this is good to go, right? /cc @dyladan

dyladan commented:
@mayurkale22 yes I think so. I didn't want to approve without your ✅
